pkg/api/handlers/projectinvitation.go

```diff
@@ -0,0 +1,332 @@
+	var (
+		code       = req.PathValue("code")
+		invitation v1.ProjectInvitation
+		thread     v1.Thread
+	)
+
+	if err := req.Get(&invitation, code); err != nil {
+		if apierrors.IsNotFound(err) {
+			return types.NewErrNotFound("invitation not found")
+		}
+		return err
+	}
+
+	if invitation.Spec.Status != string(types.ProjectInvitationStatusPending) {
+		return types.NewErrBadRequest("invitation is no longer valid")
+	}
+
+	if err := req.Get(&thread, invitation.Spec.ThreadID); err != nil {
+		return err
+	}
+
+	if thread.Spec.UserID == req.User.GetUID() {
+		return types.NewErrBadRequest("you cannot accept an invitation to your own project")
+	}
+
+	// Check if the user is already a member of the project
+	var memberships v1.ThreadAuthorizationList
+	if err := req.List(&memberships, kclient.MatchingFields{
+		"spec.threadID": invitation.Spec.ThreadID,
+		"spec.userID":   req.User.GetUID(),
+	}); err != nil {
+		return err
+	}
+
+	if len(memberships.Items) > 0 {
+		return types.NewErrBadRequest("you are already a member of this project")
+	}
+
+	// Update invitation status to accepted
+	invitation.Spec.Status = string(types.ProjectInvitationStatusAccepted)
+	if err := req.Update(&invitation); err != nil {
+		return err
+	}
+
+	// Create a new ThreadAuthorization for the user
+	threadAuth := v1.ThreadAuthorization{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: system.ThreadAuthorizationPrefix,
+			Namespace:    req.Namespace(),
+		},
+		Spec: v1.ThreadAuthorizationSpec{
+			ThreadAuthorizationManifest: types.ThreadAuthorizationManifest{
+				ThreadID: invitation.Spec.ThreadID,
+				UserID:   req.User.GetUID(),
+			},
+		},
+	}
+
+	if err := req.Create(&threadAuth); err != nil {
+		return err
+	}
+
+	var project types.Project
+	if thread.Spec.ParentThreadName != "" {
+		var parentThread *v1.Thread
+		if err := req.Get(parentThread, thread.Spec.ParentThreadName); err != nil {
+			return err
+		}
+		project = convertProject(&thread, parentThread)
+	} else {
+		project = convertProject(&thread, nil)
+	}
+
+	return req.Write(types.ProjectInvitationManifest{
+		Code:    invitation.Name,
+		Project: &project,
+		Status:  types.ProjectInvitationStatusAccepted,
+		Created: invitation.CreationTimestamp.Format(time.RFC3339),
+	})
+}
+
+func (h *ProjectInvitationHandler) RejectInvitation(req api.Context) error {
+	var (
+		code       = req.PathValue("code")
+		invitation v1.ProjectInvitation
+	)
+
+	if err := req.Get(&invitation, code); err != nil {
+		if apierrors.IsNotFound(err) {
+			return types.NewErrNotFound("invitation not found")
+		}
+		return err
+	}
+
+	if invitation.Spec.Status != string(types.ProjectInvitationStatusPending) {
+		return types.NewErrBadRequest("invitation is no longer valid")
+	}
+
+	var thread v1.Thread
+	if err := req.Get(&thread, invitation.Spec.ThreadID); err != nil {
+		return err
+	}
+
+	// Update invitation status to rejected
+	invitation.Spec.Status = string(types.ProjectInvitationStatusRejected)
```

Comment:
Is there a reason do you don't do the userID check in this function? I think it's the only one you don't do it in.

Reply:
I don't do it in GetInvitation, AcceptInvitation, and RejectInvitation. We allow all authenticated users to do this. The security comes from the code being unguessable. This is just like using an obot share link.